offchain/danaswap-api/test/Attacks/Api.purs

```diff
@@ -1,153 +1,173 @@
-  ( openPoolWrongTokenRightRedeemer
-  , openPoolWrongTokenWrongRedeemer
-  , openPoolMultipleTokens
-  , depositLiquidityWrongTokenWrongRedeemer
-  , depositLiquidityWrongTokenRightRedeemer
+  ( openPoolAttack
+  , regularOpen
+  , depositLiquidityAttack
+  , regularDeposit
   ) where
 
 import Contract.Prelude
 
-import Contract.Address (scriptHashAddress)
+import Contract.Address (getWalletAddress, scriptHashAddress)
+import Contract.BalanceTxConstraints (mustNotSpendUtxoWithOutRef) as Constraints
+import Contract.Hashing (datumHash)
 import Contract.Monad (Contract, liftContractM)
 import Contract.PlutusData (Datum(..), PlutusData(..), Redeemer(..), toData)
-import Contract.Prim.ByteArray (hexToByteArray)
 import Contract.ScriptLookups as Lookups
 import Contract.Scripts (mintingPolicyHash, validatorHash)
+import Contract.Transaction (TransactionInput)
 import Contract.TxConstraints (DatumPresence(..))
-import Contract.TxConstraints as Constraints
-import Contract.Value (mkTokenName, mpsSymbol)
+import Contract.TxConstraints (mustMintCurrencyWithRedeemer, mustMintValueWithRedeemer, mustPayToScript, mustReferenceOutput, mustSpendPubKeyOutput, mustSpendScriptOutput) as Constraints
+import Contract.Value (CurrencySymbol, TokenName, Value, mkTokenName, mpsSymbol)
 import Contract.Value as Value
-import Ctl.Utils (buildBalanceSignAndSubmitTx, getUtxos, waitForTx)
-import DanaSwap.Api (PoolId, Protocol(..), getPoolById)
+import Ctl.Utils (buildBalanceSignAndSubmitTx, buildBalanceSignAndSubmitTx', getUtxos, waitForTx)
+import DanaSwap.Api (AssetClass, PoolDatum(..), PoolId, Protocol(..), getPoolById, seedTx)
+import DanaSwap.Api (PoolId, Protocol, getPoolById)
 import DanaSwap.CborTyped (configAddressValidator)
+import Data.BigInt (BigInt, toNumber)
 import Data.BigInt as BigInt
+import Data.Int (floor)
 import Data.Map as Map
+import Math (sqrt)
 
--- TODO rework this when the real open pool gets updated
-openPoolWrongTokenWrongRedeemer :: Protocol -> Contract () PoolId
-openPoolWrongTokenWrongRedeemer (Protocol { poolAdrVal, liquidityMP, poolIdMP, configUtxo }) = do
-  poolID <- liftContractM "failed to make token name" $ mkTokenName =<< hexToByteArray "aaaa"
-  wrongID <- liftContractM "failed to make token name" $ mkTokenName =<< hexToByteArray "aabb"
-  let poolIdMph = mintingPolicyHash poolIdMP
-  poolIdCs <- liftContractM "hash was bad hex string" $ mpsSymbol poolIdMph
-  let idNft = Value.singleton poolIdCs poolID one
-  configVal <- configAddressValidator
-  configAdrUtxos <- getUtxos (scriptHashAddress $ validatorHash configVal)
-  txid <- buildBalanceSignAndSubmitTx
-    ( Lookups.mintingPolicy poolIdMP
-        <> Lookups.mintingPolicy liquidityMP
-        <> Lookups.unspentOutputs configAdrUtxos
-    )
-    ( Constraints.mustMintCurrencyWithRedeemer -- Pool id token
+-- This module provides alternative implementations
+-- for several normal API functions which take
+-- extra "attack" (someone please suggest a better word)
```

Comment:
Why the issue with `Attack`?

Reply:
This used to say "sneaky" someone suggest a better word and then you did and I used find and replace. I'll just delete the comment.

Reply:
Well to give credit to where it's due it was actually @marijanp that came up with the name. I just liked it a lot 😄